prow/githuboauth/githuboauth.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package githuboauth
+
+import (
+	"crypto/subtle"
+	"encoding/hex"
+	"fmt"
+	"net/http"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+	"golang.org/x/net/xsrftoken"
+	"golang.org/x/oauth2"
+
+	"k8s.io/test-infra/prow/config"
+)
+
+const (
+	oauthSessionCookie = "oauth-session"
+	stateKey           = "state"
+)
+
+type OAuthClient interface {
+	Exchange(ctx context.Context, code string) (*oauth2.Token, error)
+	AuthCodeURL(state string, opts ...oauth2.AuthCodeOption) string
+}
+
+// GithubOAuth Agent represents an agent that takes care Github authentication process such as handles
+// login request from users or handles redirection from Github OAuth server.
+type GithubOAuthAgent struct {
+	gc     *config.GithubOAuthConfig
+	logger *logrus.Entry
+}
+
+// Returns new GithubOAUth Agent.
+func NewGithubOAuthAgent(config *config.GithubOAuthConfig, logger *logrus.Entry) *GithubOAuthAgent {
+	return &GithubOAuthAgent{
+		gc:     config,
+		logger: logger,
+	}
+}
+
+// HandleLogin handles Github login request from front-end. It starts a new git oauth session and
+// redirect user to Github OAuth end-point for authentication.
+func (ga *GithubOAuthAgent) HandleLogin(client OAuthClient) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		stateToken := xsrftoken.Generate(ga.gc.ClientSecret, "", "")
+		state := hex.EncodeToString([]byte(stateToken))
+		oauthSession, err := ga.gc.CookieStore.New(r, oauthSessionCookie)
+		if err != nil {
+			ga.serverError(w, "Creating New OAuth session", err)
+			return
+		}
+		oauthSession.Options.MaxAge = 10 * 60
+
+		if err != nil {
+			ga.serverError(w, "Create new oauth session", err)
+			return
+		}
```

Comment:
This error check is a duplicate of the one above and can be removed.

Reply:
done